clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6203,7 +6217,102 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    // Generate forward declaration for free function.
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    if (K.SyclKernel->getLanguageLinkage() == CLanguageLinkage)
+      O << "extern \"C\" ";
+    std::string ParmList;
+    bool FirstParam = true;
+    for (ParmVarDecl *Param : K.SyclKernel->parameters()) {
+      if (FirstParam)
+        FirstParam = false;
+      else
+        ParmList += ", ";
+      ParmList += Param->getType().getCanonicalType().getAsString();
+    }
+    FunctionTemplateDecl *FTD = K.SyclKernel->getPrimaryTemplate();
+    Policy.SuppressDefinition = true;
+    Policy.PolishForDeclaration = true;
+    if (FTD) {
+      FTD->print(O, Policy);
+    } else {
+      K.SyclKernel->print(O, Policy);
+    }
+    O << ";\n";
+
+    // Generate a shim function that returns the address of the free function.
+    O << "static constexpr auto __sycl_shim" << ShimCounter << "() {\n";
+    O << "  return (void (*)(" << ParmList << "))"
```

Comment:
Sorry I didn't bring this up earlier but why is this explicit cast actually required? The cast happens implicitly work - https://godbolt.org/z/3qcxrMxPG

Reply:
The cast is not always needed. In the test ff_2 is an overloaded function with three or four parameters. A cast is necessary to distinguish between them. Rather than trying to determine whether or not a cast is needed, which is likely to be difficult, we always use a cast as it does no harm and keeps the implementation simpler.

Reply:
Thanks for explaining!